sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala

```diff
@@ -142,7 +142,9 @@ object RowEncoder {
     case CalendarIntervalType => dt
     case TimestampType => ObjectType(classOf[java.sql.Timestamp])
     case DateType => ObjectType(classOf[java.sql.Date])
-    case _: DecimalType => ObjectType(classOf[java.math.BigDecimal])
+    // In order to support both Decimal and java BigDecimal in external row, we make this
+    // as java.lang.Object.
+    case _: DecimalType => ObjectType(classOf[java.lang.Object])
```

Comment:
`externalDataTypeFor` is also used in `constructorFor`, I think we should create a `externalDataTypeForInput` that can handle decimal type specially and use it in `extractorFor`.